### ISSUE TITLE
Weird history from ACME merge a few months ago

### ISSUE THREAD

**Issue by billsacks** [ESMCI/cime#2177]
Nothing needs to be done about this - indeed, I doubt anything reasonable could be done about it at this point. But I wanted to point it out so that we can hopefully identify the cause and prevent oddness like this in the future.

If I do a git blame of `scripts/fortran_unit_testing/run_tests.py` (I suspect I'd see the same for other files; I just noticed it for that one), I see most of the lines labeled as

```
^587cf5950 (Azamat Mametjanov 2017-05-20 01:34:50 +0000
```

The leading caret apparently means that there is no parent of this commit. That commit was merged to master in the big acme merge:

```
commit 0c1b306aa
Merge: 4f201fe2d 2df9636fa
Author: James Foucar <jgfouca@sandia.gov>
Date:   Mon Oct 23 11:39:42 2017 -0600

    Merge pull request #1970 from ESMCI/jgfouca/branch-for-acme-split-10-17-2017-pr

    Bring in E3SM cime changes for past 3 months
```

Looking at the history along the branch that resulted in that ACME merge, it looks like it's common for the commits to be parentless according to the git history (presumably because they're extracted out of the ACME history). But I'm quite sure that Azamat was not the original author of most of the lines in run_tests.py, as it now appears according to a `git blame`.

@jgfouca or others: Any ideas how we can maintain proper history of files in the future (via git blame, etc.)?

**Comment by jgfouca** [comment]
@billsacks I think we are at the mercy of git subtree unfortunately.

**Comment by billsacks** [comment]
@jgfouca Spot-checking a few other files, it looks like this problem is wide-spread. For example, `git blame src/drivers/mct/shr/seq_flds_mod.F90` makes it look like that file was written almost entirely by Pat Worley, and `git blame src/drivers/mct/main/cime_comp_mod.F90` makes it look like that file was written almost entirely by Steve Ghan. This makes `git blame` and similar history tools all but useless. It's hard for me to believe that this wide-spread history mess-up is an unavoidable result of using git subtree. If there is truly no way to avoid this with git subtree then I'd oppose further uses of git subtree to bring changes back into cime, since I feel that these history mess-ups are a pretty big deal.

If you can't figure out what aspect of ACME's subtree-based workflow is responsible for this – and thus, what could be changed to prevent it – then could you please describe this workflow in detail so that I or others can help you figure out if anything can be done to prevent this problem in the future?

**Comment by jgfouca** [comment]
@billsacks , I've been against subtree all along as it opens the pandora's box of history-rewriting operations in Git which, in my experience, leads to endless amounts of pain and confusion. Integrating ESMCI within ACME via submodule is the path I preferred but I was overridden by higher-up people in ACME. Their argument is that it would make it difficult for ACME devs to make config changes and I have to admit that's true.

Anyways, our usage of subtree is pretty standard as far as I can tell, just basic subtree splits and pulls. You can see the details in scripts/lib/acme_cime_mgmt.py; the two main driver routines are at the bottom of the file. I agree that if this is causing damage to the history of the repo, then maybe we need some other solution. @rljacob any thoughts here?

**Comment by billsacks** [comment]
Thanks for your thoughts @jgfouca 

First, big caveat: I know just enough to be dangerous with git subtrees.

That said: It looks to me like the problem is that you're regularly using `git subtree split`. My understanding is that you're only supposed to use that command in the scenario where you have a big repo and you want to split it up to begin managing it via subtrees. (In this case, it would be as if cime started out as part of ACME, and then we decided that we wanted a separate cime repository). Once you already have a separate repository for the subrtree, my understanding is that you're supposed to use `git subtree push`. I'd be interested to see an experiment where you change `git subtree split` to `git subtree push` (possibly changing other command arguments) and seeing if that cleans up the history.

**Comment by billsacks** [comment]
By the way: For those who stumble upon this thread looking for a workaround to get the correct `git blame` output: The way I have done it is:

1. Do a `git blame` on the file; find the commit referenced on most lines with a leading `^`. For example, for `run_tests.py`, this is `^587cf5950`

2. Determine the commit on master in which the above commit was merged. I have used https://github.com/mhagger/git-when-merged for this purpose (on a mac, this can be installed with `brew install git-when-merged`). In this case, I get:

```
[roo2:~/cime]$ git when-merged 587cf5950
refs/heads/master                      0c1b306aab677db75824b0315b9a8471a9225fab
```

3. Determine the parent of the above commit on master:

```
[roo2:~/cime]$ git show 0c1b306aab677db75824b0315b9a8471a9225fab^
commit 4f201fe2d
```

4. Do a `git blame` from that commit: e.g., `git blame 4f201fe2d -- scripts/fortran_unit_testing/run_tests.py`

This is doable when necessary, but becomes tiresome quickly :-)

**Comment by jgfouca** [comment]
@billsacks , it looks like you're right, we weren't doing the splits correctly. I wonder if adding the --rejoin flag would have made it equivalent to subtree push?

**Comment by jgfouca** [comment]
From the documentation:

```
       push
           Does a split (see below) using the <prefix> supplied and then does
           a git push to push the result to the repository and ref. This can
           be used to push your subtree to different branches of the remote
           repository.
```

So if push is just doing a split under the hood, then maybe the history will still be messed up?

**Comment by jgfouca** [comment]
This post:

https://stackoverflow.com/questions/12801936/git-subtree-push-changes-from-an-cloned-repo

... seems to indicate that adding --rejoin is necessary to get the same behavior you would get from subtree push. I'm going to try adding it to my script and see what happens.

**Comment by billsacks** [comment]
I am unable to reproduce the behavior I'm seeing in cime (i.e., completely rewritten history) with either a `git subtree push` or a `git subtree split` using a simple test case (roughly following the tutorial in https://medium.com/@porteneuve/mastering-git-subtrees-943d29a798ec). I no longer think that the problem is the difference between `push` and `split`. I tried to reproduce the `subtree split` process that's done by `acme_cime_mgmt.py` from my ACME clone, but ran into problems and gave up. If you still can't figure it out, it would be helpful if you could distill the process into a set of git commands that could be run from the command line that reproduce what's done for an acme cime split - then we could go from there.

**Comment by jgfouca** [comment]
@billsacks , thanks for helping with this. I was playing around with this last night and 'subtree push' and 'split --rejoin' definitely behave differently than what I was doing before. In many cases, I ended up with a history that was completely detached and therefore un-merge-able.

I have a PR coming soon that cleans up acme_cime_mgmt a bit and prints the key commands, that should help.

@jedbrown @douglasjacobsen , do either of you know much about subtree?

**Comment by douglasjacobsen** [comment]
I do, what do you want to know other than how messy they are?

**Comment by jgfouca** [comment]
@douglasjacobsen , you're telling me ;) !

I wish I could ask a specific question, but I'm confused to the extent that I don't even know what to ask. I fear, due mostly to my ignorance, that we've already gone down a bad path in our subtree usage and I don't know how to recover. A simple git subtree pull/push --squash model probably would have been best, but it's too late for that unfortunately.

Let me see if I can explain the current algorithm:

To split (move ACME cime changes to ESMCI):
* Update current master branch from origin (ACME) and esmci remotes
* Tag current ACME commit with a "split" tag
* Query the most recent "merge" tag from esmci
* Query the preview "split" tag from ACME
* Do the split:
    git subtree split $old_split_tag.. --prefix=cime --onto=$merge_tag --ignore-joins -b $new_split_branch
* Merge esmci/master into $new_split_branch, this step usually involves fixing conflicts
* Push $new_split_branch as a PR for ESMCI

To merge (import ESMCI cime changes into ACME):
* Update current master branch from origin (ACME) and esmci remotes
* Query old "merge" tag from ESMCI
* Make new "merge" tag for ESMCI for current esmci/master
* Make a new branch from acme/master
* Do a subtree pull:
    git subtree pull --prefix=cime $esmci_remote master
* Fix conflicts from prior step
* Push current branch as a PR for ACME

**Comment by jgfouca** [comment]
I think the problem is that I was trying to micromanage too much with those tags.

**Comment by douglasjacobsen** [comment]
So, just looking at the history of cime / ACME a bit, the issue related to @billsacks original comment doesn't appear to be related to the use of subtrees. In both CIME and ACME that commit is an orphan. Additionally, there are multiple commits that are identical in ACME (though only one in CIME). That commit was made in May, so I'd be surprised if you could figure out how it was created, but that's what you need to find out.

Whether or not you started with `--squash` on subtrees, it's never too late to change. The only thing that seems a bit odd to me is the use of `--ignore-joins`.

However, I think @billsacks problem is related to strictly relying on `git blame`. For example, one could easily use `git log --oneline --decorate --graph <path_to_file>` to see all commits that touch that file, and then use `git blame` after checking out a different version (for example 3646998). It seems like the main issue is that @amametjanov's commit is in the first parent history, so it ignores the other side of the merge on 2ea468c, but if you do:

`git checkout 2ea468c^2 && git blame scripts/fortran_unit_tests/run_tests.py` I think that will show what you're looking for.

**Comment by billsacks** [comment]
Thanks for your additional thoughts, @douglasjacobsen and @jgfouca 

@douglasjacobsen from what I can tell, that commit from @amametjanov was originally a clean commit, in that it is shown as just touching a few files:

```
$ git show --name-status 4a7214dc3
commit 4a7214dc3
Author: Azamat Mametjanov <azamat@mcs.anl.gov>
Date:   Sat May 20 01:34:50 2017 +0000

    Define threading env-vars only for threaded builds

M       cime/config/acme/machines/config_machines.xml
M       cime/scripts/lib/CIME/XML/env_mach_specific.py
M       cime/scripts/lib/CIME/case.py
M       cime/src/drivers/mct/cime_config/config_component.xml
```

The duplicate commit in ACME that is an orphan (587cf5950) seems to be a symptom, not a cause of the problem: that commit seems to have originated from @jgfouca 's pulling cime back into ACME after the problematic subtree split from ACME to cime had already happened. So I still feel like the problem originated in the subtree splitting, not in the ACME history initially.

@jgfouca do you feel like you have things figured out at this point, or could you still use more help?

**Comment by jgfouca** [comment]
@billsacks , no, in fact, I'm worse off than before. I'm finding now that my original implementation of acme_cime_split is no longer working. It's producing a history that has no common commits with esmci master.

**Comment by douglasjacobsen** [comment]
@billsacks Sure. I'd agree that it's a symptom, but it's what is causing your weird behavior. I think the cause is likely weird subtree split stuff that's been happening.

@jgfouca I don't have a lot of time to mess with this, but if you want to post example commands you're using to combine historys, if I get time I can look them over to see if I can help some.

**Comment by billsacks** [comment]
@jgfouca shoot, that doesn't sound good. This is getting a bit out of my depth, so I'm not sure how much help I'll be, but I'll continue to watch this thread if you learn anything more.

**Comment by jedbrown** [comment]
Just catching up, I would think `--squash` would not be desirable because it would lose commit identity when bringing ESMCI changes into ACME. I also think the `--ignore-joins` looks odd. It has been a few years since I've used subtree for two-way (versus one-way) integration, but I'd be curious to see a test case.

**Comment by douglasjacobsen** [comment]
@jedbrown Your comment about `--squash` is definitely correct. I think it'd affect it both ways though. ESMCI wouldn't know who in ACME made changes, and ACME wouldn't know who in ESMCI made changes. It'd definitely make debugging commit histories a lot harder, and things like `git blame` would work even worse than they currently do.

**Comment by jgfouca** [comment]
@billsacks , I recently did another E3SM split and it seemed to work pretty well. Is there any way to verify that the history was not corrupted?

**Comment by rljacob** [comment]
Re: --ignore-joins, according to my notes, for the command
`git subtree split -P <external_subdir> -b <external_branchname>` 
 if the external_subdir has received more then one merge from the external, you have to add --ignore-joins to the above or you get an error.

**Comment by billsacks** [comment]
History was still corrupted for at least some files. As a first guess (unconfirmed), I'm thinking that the files with messed up history (at least with respect to `git blame`) are those that changed in the subtree split/merge. I spot-checked a few of these:

```
[roo2:~/cime]$ git diff --name-status 18358ce1c^ 18358ce1c
M       config/acme/allactive/config_pesall.xml
M       config/acme/config_grids.xml
M       config/acme/machines/config_batch.xml
M       config/acme/machines/config_compilers.xml
M       config/acme/machines/config_machines.xml
A       config/acme/testmods_dirs/allactive/mach_mods/shell_commands
M       scripts/lib/CIME/wait_for_tests.py
M       scripts/lib/update_acme_tests.py
M       src/drivers/mct/main/prep_ocn_mod.F90
M       src/drivers/mct/main/seq_rest_mod.F90
```

and at least some of them still have a `git blame` that makes it look like they were newly created by the person who made a change. For example, from master, see: `git blame scripts/lib/update_acme_tests.py`

**Comment by billsacks** [comment]
For what it's worth: I just tested what would have happened if @jgfouca had merged the `jgfouca/branch-for-acme-split-2018-01-09` branch to master using `--squash`. From spot-checking `git blame scripts/lib/update_acme_tests.py`, it looks like this would fix the symptom of the problem from the point of view of cime history. Same is true for the three files mentioned above from the previous acme merge.

I understand that `--squash` has its own issues. The real solution is probably to change how the acme split branches are created in the first place, in some way. But from my perspective, until the underlying issue can be resolved, the end result of the `--squash` is preferable to the end result of the non-squash. So I'd vote for doing that as an interim solution.

To address the issues with squash (losing history), I'd consider dividing cime commits in E3SM into two categories:

(1) Changes that are only of interest to E3SM, such as in config/acme. These seem relatively safe to combine into a single PR and then squash to master.

(2) Changes that affect other cime users (i.e., affect cesm as well as e3sm). For these, I'd argue that they should ideally each appear in their own PR to cime anyway. In that case, the negative implications of using `--squash` would be less: you could even set the `--author` flag when committing to achieve proper attribution on each change.

**Comment by jgfouca** [comment]
Thanks @billsacks , I agree that squash is probably best, at least for ACME->ESMCI.

**Comment by billsacks** [comment]
@jgfouca I'm okay with closing this if you are.

**Comment by jgfouca** [comment]
Thanks, I forgot to close.

**Comment by billsacks** [comment]
@jgfouca today's discussions led me to get re-obsessed with what's going on here. I know we have a workaround (always bringing in your subtree changes with a `--squash` merge to cime master), but I'm not completely happy with that solution because:

1. We're losing the history of who in ACME actually made various changes

2. I'm worried that someone is going to forget the squash sometime, and we'll be back to having a hosed history

So I spent a little more time on this. And I think I'm closer to understanding what aspect of your subtree workflow is causing the problem: I think it's related to your use of `old_split_tag..` in this command:

```python
    run_cmd_no_fail("git subtree split {}.. --prefix=cime --onto={} -b {}".\
                        format(old_split_tag, merge_tag, subtree_branch), verbose=True)
```

I ran through portions of the example in this tutorial: https://medium.com/@porteneuve/mastering-git-subtrees-943d29a798ec

When I introduced a change to the subtree and then did `git subtree split -P vendor/plugins/demo/ b880d24.. -b newsplit` (where b880d24 was the commit in the container in which the subtree was squashed - i.e., the commit with message, "Squashed 'vendor/plugins/demo/' content from commit fe64799"), I was able to reproduce the behavior we're seeing, where it looks like the file I touched was completely rewritten, without a parent. If I removed the `b880d24..` from the command, things worked as I'd expect (e.g., when doing `git blame`).

This is also addressed in this stackoverflow post: https://stackoverflow.com/questions/5760331/git-subtree-is-not-retaining-history-so-i-cannot-push-subtree-changes-how-can-i  -- specifically, see footnote 1 of the accepted answer.

I'm not sure what purpose your `old_split_tag..` is serving in your workflow, but I wonder if it would make sense to get rid of that? I'm getting the impression that `git subtree` misbehaves when you try to be explicit like that. I'm also getting the impression that you're trying to be a bit more explicit about things rather than relying on `git subtree`'s implicit behaviors. I can see why you'd want to do that, but I wonder if you'd be better served in that case doing things completely manually rather than trying to use `git subtree` at all. (e.g., would things work better if you used the "manually" approach from https://medium.com/@porteneuve/mastering-git-subtrees-943d29a798ec ?).

**Comment by billsacks** [comment]
Here's how to reproduce the behavior we're seeing:

Click "Download the example repos" from https://medium.com/@porteneuve/mastering-git-subtrees-943d29a798ec and unzip it

```
cd git-subs/main
git remote add plugin ../remotes/plugin/
git subtree add --prefix=vendor/plugins/demo --squash plugin master
echo '// Now super fast' >> vendor/plugins/demo/lib/index.js
git commit -am "[To backport] Faster plugin"
git log  # find the commit that says "Squashed 'vendor/plugins/demo/' content from commit fe64799"; in my case, that commit's sha-1 is 9804fa6
git subtree split -P vendor/plugins/demo/ 9804fa6.. -b newsplit  # replace 9804fa6 with the commit you found
git checkout newsplit
git blame lib/index.js
```

Then redo this from scratch, but without `9804fa6..` and observe the difference.

**Comment by jgfouca** [comment]
@billsacks , thanks, this is great information. I'll see if I can improve the script.

**Comment by jgfouca** [comment]
Fixed by #2314
